src/qutip_tensornetwork/core/data/network.py

```diff
@@ -0,0 +1,627 @@
+from tensornetwork.visualization.graphviz import to_graphviz
+from tensornetwork.network_components import AbstractNode, Node, Edge
+from tensornetwork.network_components import CopyNode
+from tensornetwork.network_operations import (get_all_nodes, reachable,
+                                              get_all_edges
+                                             )
+from tensornetwork.network_operations import get_subgraph_dangling
+from tensornetwork.contractors import greedy
+
+__all__ = ['Network']
+
+class Network(qutip.core.data.Data):
+    """Represents arbitrary quantum objects as tensor networks.
+
+    Tensor networks will be composed of ``tensornetwork.Nodes`` and
+    ``tensornetwork.Edges``. Nodes represent n-dimensional arrays whose
+    dimensions are connected with edges to other arrays. An edge represents the
+    contraction that needs to be carried out between two nodes. The contraction
+    of nodes is done in a `lazy` way, which means that is only carried out when
+    necessary or explicitly required with the contract method.
+
+    Edges can be either dangling (one of its ends is not connected to any node)
+    or fully connected. To represents arbitrary quantum objects, we employ the
+    dangling edges ``out_edges`` and ``in_edges``. Considered as a matrix, the
+    dangling ``out_edges`` and ``in_edges`` represent the rows and columns of
+    the matrix respectively.
+
+    Attributes
+    ----------
+    nodes : set of Nodes
+        Nodes that belong to the Network. These can either be reachable from
+        in_edges and out_edges or scalar nodes, in which case they have no
+        edges.
+
+    out_edges : list of Edges
+        List of ``Edges`` to be used. When the network is considered as a
+        matrix, these edges represent the rows.
+
+    in_edges : list of Edges
+        List of ``Edges`` to be used. When the network is considered as a
+        matrix, these edges represent the columns.
+
+    dims : list of int
+        Dimension of the system as a list of lists. dims[0] represents the
+        out dimensions whereas dims[1] represents the in dimension.
+
+    shape : tuple of int
+        Shape that the matrix would have if the network is represented with a
+        matrix.
+
+    Notes
+    -----
+    Most of the operations and logic of this class has been derived from
+    ``tensornetwork.quantum.QuOperator``. However, this class is not compatible with
+    ``QuOperator`` but we provide a method, ``as_quoperator()`` that returns a view
+    of this class as a ``QuOperator``.
+    """
+
+    def __init__(self, out_edges, in_edges, nodes = None, copy = True):
+        """Creates a new ``Network`` from a tensor network.
+        This encapsulates an existing tensor network, interpreting it as a linear
+        operator.
+
+        The network is checked for consistency: All dangling edges must either be
+        in `out_edges`, `in_edges`, or `ignore_edges`.
+
+        Parameters
```

Comment:
It's more common and useful to document the parameters for `__init__` in the class docstring (since from user code one calls `MyClass(x, y, z)` and not `MyClass.__init__(...)`.